[ConstraintSolver] When ranking bindings prefer type vars with literal bindings

xedin opened the pull request:
Change the potential binding ranking logic to prefer type variables
which don't have other type variables related to them, but have literal
bindings, over the ones that do have other type variables.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSolver] When ranking bindings prefer type vars with literal bindings

Change the potential binding ranking logic to prefer type variables
which don't have other type variables related to them, but have literal
bindings, over the ones that do have other type variables.

xedin commented:
@swift-ci please smoke test

slavapestov commented:
What does this give us?

xedin commented:
@slavapestov Sometimes we pick something for "best" bindings which isn't really that, e.g. when I remove contraction of `BindParam` constraint this change makes sure that we don't pick type variable from either side of `bind param` constraint unless one of the sides is already bound or there is no better literal bindings. That would allow us to remove at least one flag from `PotentialBindings` bindings.

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test Linux platform

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@rudkx Looks like it improved one of the slow perf tests too by applying literals earlier \o/

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@rudkx I've added couple more steps to the test which is now much faster and moved it to perf/fast.

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

slavapestov commented:
Awesome!

Although that test case should be diagnosed as invalid :( We don't allow single-element tuples as values anymore.

xedin commented:
@slavapestov This is intentional, some of the code exhibits exponential behavior when we can't easily find a solution or there is no solution e.g. `validation-test/Sema/type_checker_perf/slow/rdar19737632.swift`
```swift
let a = "a"
let b = "b"
let c = 42
_ = "a=" + a + ";b=" + b + ";c=" + c
```

Is currently "too complex" because it can't find a solution with favorite types and solver goes on exploring generics forever...

xedin committed:
[ConstraintSolver] Attempt literal bindings earlier

If the best bindings we could get on the current step are bindings
to literal type without any other type variable involved, let's try
to attempt them right away, that should help to prune search space.

xedin committed:
[ConstraintSolver] NFC: Refactor `PontentialBindings` to reference type variable they belong to

xedin commented:
@rudkx I've added commit which removes restrictions for early attempting of literal bindings and refactored `determineBestBindings` to use Optional instead of creating empty bindings. Let's see if everything is going to work together :)

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

rudkx reviewed (approved):
LGTM. I had tried the same change to CSSolver.cpp not that long ago but I think something failed. Maybe it also requires the other change to type check one of the expressions or perhaps something else has changed in the meantime.

xedin commented:
@swift-ci please smoke test and merge

xedin commented:
@swift-ci please smoke test and merge
